File: packages/provide-uterm-csharp/src/Provide.Uterm/Embed/Hub.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

namespace Provide.Uterm.Embed;

/// <summary>
/// In-process session factory. Hosts (e.g. TWX30) call
/// <c>await hub.CreateSessionAsync(options)</c> — no CLI or loopback HTTP required.
/// </summary>
public sealed class EmbedHub : IEmbedHub
{
    private readonly Dictionary<string, IUtermSession> _sessions = new(StringComparer.Ordinal);
    private readonly object _gate = new();
    private int _seq;

    public IReadOnlyCollection<string> SessionIds
    {
        get
        {
            lock (_gate)
            {
                return _sessions.Keys.ToArray();
            }
        }
    }

    public Task<IUtermSession> CreateSessionAsync(EmbedSessionOptions? options = null, CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        options ??= new EmbedSessionOptions();
        var id = string.IsNullOrEmpty(options.SessionId)
            ? "embed-" + Interlocked.Increment(ref _seq).ToString("x", System.Globalization.CultureInfo.InvariantCulture)
            : options.SessionId;

        var session = new UtermSession(id, options);
        lock (_gate)
        {
            if (_sessions.ContainsKey(id))
            {
                throw new InvalidOperationException("session already exists: " + id);
            }

            _sessions[id] = session;
        }

        return Task.FromResult<IUtermSession>(session);
    }

    public IUtermSession? GetSession(string sessionId)
    {
        lock (_gate)
        {
            return _sessions.TryGetValue(sessionId, out var s) ? s : null;
        }
    }

    /// <summary>Remove a disposed session from the registry.</summary>
    public bool RemoveSession(string sessionId)
    {
        lock (_gate)
        {
            return _sessions.Remove(sessionId);
        }
    }
}

File: packages/provide-uterm/src/provide/uterm/embed/hub.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""In-process embed session factory."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from provide.uterm.embed.session import EmbedSession

if TYPE_CHECKING:
    from collections.abc import Mapping

    from provide.uterm.embed.session import ByteInterceptor
    from provide.uterm.embed.types import TelnetPolicy

__all__ = ["EmbedHub"]


class EmbedHub:
    """In-process session factory."""

    def __init__(self) -> None:
        self._sessions: dict[str, EmbedSession] = {}
        self._seq = 0

    @property
    def session_ids(self) -> list[str]:
        return list(self._sessions)

    async def create_session(
        self,
        *,
        session_id: str | None = None,
        interceptor: ByteInterceptor | None = None,
        telnet_policy: TelnetPolicy | None = None,
        services: Mapping[str, Any] | None = None,
    ) -> EmbedSession:
        if not session_id:
            self._seq += 1
            session_id = f"embed-{self._seq:x}"
        if session_id in self._sessions:
            raise RuntimeError(f"session already exists: {session_id}")
        sess = EmbedSession(
            session_id,
            interceptor=interceptor,
            telnet_policy=telnet_policy,
            services=services,
        )
        self._sessions[session_id] = sess
        return sess

    def get_session(self, session_id: str) -> EmbedSession | None:
        return self._sessions.get(session_id)

    def remove_session(self, session_id: str) -> None:
        self._sessions.pop(session_id, None)
